controllers/services_controller.go

```diff
@@ -0,0 +1,181 @@
+			Type:            v1.ServiceTypeClusterIP,
+		},
+	}
+}
+
+// Annotation to generate RBAC roles to read and modify services
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;watch;list;create;update;delete
+
+func CreateServiceController(mgr ctrl.Manager) (*serviceReconciler, error) {
+	return newServiceReconciler(mgr)
+}
+
+func (r *serviceReconciler) Start(ctx context.Context, mgr ctrl.Manager) error {
+	err := r.createMetricsService(ctx, mgr.GetLogger())
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return fmt.Errorf("error start serviceReconciler: %w", err)
+	}
+
+	return r.setupController(mgr)
+}
+
+func (r *serviceReconciler) setServiceOwnerReference(service *v1.Service) error {
+	return controllerutil.SetOwnerReference(r.deployment, service, r.client.Scheme())
+}
+
+func (r *serviceReconciler) createMetricsService(ctx context.Context, logger logr.Logger) error {
+	service := ServiceObject(r.serviceNamespace)
+	err := r.setServiceOwnerReference(service)
+	if err != nil {
+		return fmt.Errorf("error setting owner reference: %w", err)
+	}
+	return r.client.Create(ctx, service)
+}
+
+func (r *serviceReconciler) setupController(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("service-controller").
+		For(&v1.Service{}).
+		Complete(r)
+}
+
+// serviceReconciler reconciles the required services in the operator's namespace
+type serviceReconciler struct {
+	client           client.Client
+	apiReader        client.Reader
+	log              logr.Logger
+	serviceNamespace string
+	deployment       *apps.Deployment
+}
+
+func (r *serviceReconciler) setDeploymentField() error {
+	objKey := client.ObjectKey{Namespace: r.serviceNamespace, Name: OperatorName}
+	var deployment apps.Deployment
+	err := r.apiReader.Get(context.TODO(), objKey, &deployment)
+	if err != nil {
+		return fmt.Errorf("setDeploymentField, get deployment: %w", err)
+	}
+	r.deployment = &deployment
+	return nil
+}
+
+func newServiceReconciler(mgr ctrl.Manager) (*serviceReconciler, error) {
+	logger := ctrl.Log.WithName("controllers").WithName("Resources")
+	namespace, err := getOperatorNamespace(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	reconciler := &serviceReconciler{
+		client:           mgr.GetClient(),
+		apiReader:        mgr.GetAPIReader(),
+		log:              logger,
+		serviceNamespace: namespace,
+	}
+
+	err = reconciler.setDeploymentField()
+	if err != nil {
+		return nil, err
+	}
+
+	return reconciler, nil
+}
+
+func getOperatorNamespace(logger logr.Logger) (string, error) {
+	nsBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		return "", err
+	}
+	ns := strings.TrimSpace(string(nsBytes))
+	logger.Info("Found namespace", "Namespace", ns)
+	return ns, nil
+}
+
+func (r *serviceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	if req.Name != MetricsServiceName || req.Namespace != r.serviceNamespace {
```

Comment:
Can you move this check into a predicate inside `setupController()` ?
For example:
```go
For(&v1.Service{}, builder.WithPredicates(predicate.NewPredicateFuncs(
	func(object client.Object) bool {
		return object.GetName() == MetricsServiceName && object.GetNamespace() == r.serviceNamespace
	}))).
```